analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs

```diff
@@ -26,3 +26,22 @@ public abstract class SonarAnalysisContextBase
 {
     public abstract bool TryGetValue<TValue>(SourceText text, SourceTextValueProvider<TValue> valueProvider, out TValue value);
 }
+
+public abstract class SonarAnalysisContextBase<TContext> : SonarAnalysisContextBase
+{
+    public abstract SyntaxTree Tree { get; }
+    public abstract Compilation Compilation { get; }
+    public abstract AnalyzerOptions Options { get; }
+
+    public SonarAnalysisContext AnalysisContext { get; }
```

Comment:
It would be useful to describe in the PR why is the hierarchy needed and how will they be used. I think it's easy to miss something during code review since the context is missing.

e.g. This property is of type `SonarAnalysisContext` that derives from `SonarAnalysisContextBase`. So to properly construct an instance you need another instance :)

I'm missing the general picture.

Reply:
```mermaid
graph TB;
SonarAnalysisContextBase--> SonarAnalysisContext ;
SonarAnalysisContextBaseOfTContext --> SonarReportingContextBaseOfTContext;
SonarAnalysisContextBase --> SonarAnalysisContextBaseOfTContext;
SonarAnalysisContextBase --> SonarParametrizedAnalysisContext;
SonarCompilationReportingContextBaseOfTContext --> SonarSymbolAnalysisContext;
SonarTreeReportingContextBaseOfTContext --> SonarSyntaxNodeAnalysisContext;
SonarReportingContextBaseOfTContext --> SonarTreeReportingContextBaseOfTContext;
SonarReportingContextBaseOfTContext --> SonarCompilationReportingContextBaseOfTContext;
SonarTreeReportingContextBaseOfTContext --> SonarCodeBlockAnalysisContext;
SonarAnalysisContextBase --> SonarCodeBlockStartAnalysisContext;
SonarCompilationReportingContextBaseOfTContext --> SonarCompilationAnalysisContext;
SonarAnalysisContextBase --> SonarCompilationStartAnalysisContext;
SonarTreeReportingContextBaseOfTContext --> SonarSyntaxTreeAnalysisContext;
```

Reply:
I've added a diagram to the #6532 description

Reply:
There's a problem that some of the analysis-time contexts (classes inheriting from `SonarReportingContextBase` in our case) do not have access to the stuff from `SonarAnalysisContext` and they often need.

So from now on, each reporting context will hold a reference to the root `SonarAnalysisContext`.

Reply:
In derived classes, how do you know if you need to call `TryGetValue` from current class or from `SonarAnalysisContext`?

Reply:
As discussed, the intended inheritance was redesigned - will be in PR 20

Reply:
@andrei-epure-sonarsource Please see #6532 for the up-to-date targeted diagram. With the latest changed ( #6635 ) and renaming of the reporting contexts (bottom leaves) to `*ReportingContext`, there's already a separation between those that report and those that (mainly) register. So they are in distinct group names. I think it's good enough now.

Reply:
@pavel-mikula-sonarsource regarding the base classes, I find it very confusing to have two totally decoupled classes called `SonarAnalysisContext` and `SonarAnalysisContextBase`...

Reply:
It's unfortunate that Roslyn has every class in that area called `*AnalysisContext`, while they have quite different use scenarios. 
![image](https://user-images.githubusercontent.com/57188685/212915573-c87d7360-552b-4a1b-b6cb-cc0148366c0d.png)

They are not decoupled, `SonarAnalysisContextBase` actually heavily depends on `SonarAnalysisContext` and needs it to be constructed. The problem is the `Analysis` word there that I don't know how to properly replace. The `RootContext` doesn't say it's related to analysis, it more replaces the `Base` part. And we have many other contexts around the codebase. 

This effort already took too much time and we need to stop somewhere. As I believe we're in a good shape now, I wouldn't go further for now.

Reply:
I agree to stop refactoring. 

I believe that, now that the context is fresh (pun intended), having some documentation and explanation in the docs section would be helpful, especially with the enlarging of the team. 

And I feel that these low level classes won't change much, so the risk of the docs getting out of date is small. 

WDYT? At least the class structure and how we should be using them correctly